### PR TITLE
[prometheus] major refactor

### PR DESCRIPTION
**Pull request by zanac1986**
#### What this PR does / why we need it

Major refactoring several syntax and fixing mis-alignment across chart components

#### Which issue this PR fixes

- Empty values (e.g. `[]`, `{}`) shall omit.
```
# from
{{- if .resources }}
resources:
  {{- toYaml .resource | nindent 2 }}
{{- end }}
# to
{{- with .resources }}
resources:
  {{- toYaml . | nindent 2 }}
{{- end }}
```
- Misuse truncate empty spaces + indention.
```
# from
metadata:
  ...
  name: foo
{{ include "foo.namespace" $ | indent 2 }}
...
# to
metadata:
  ...
  name: foo
  {{- include "foo.namespace" $ | nindent 2 }}
...
```
- Use scope. Note that we have to change all `include` and `template` as root based (`.` -> `$`).
```
# from
apiVersion: v1
kind: Service
metadata:
{{- if .Values.foo.a.super.super.long.miles.away.annotations }}
  annotations:
{{ toYaml .Values.foo.a.super.super.long.miles.away.annotations | indent 4 }}
{{- end }}
  labels:
    {{- include "foo.labels" . | nindent 4 }}
{{- if .Values.foo.a.super.super.long.miles.away.labels }}
{{ toYaml .Values.foo.a.super.super.long.miles.away.labels | indent 4 }}
{{- end }}
  name: {{ template "foo.fullname" . }}-headless
{{ include "foo.namespace" . | indent 2 }}
spec:
  clusterIP: None
  ports:
    - name: http
      port: {{ .Values.foo.a.super.super.long.miles.away.servicePort }}
      protocol: TCP
      targetPort: 9093
{{- if .Values.foo.a.super.super.long.miles.away.enableMeshPeer }}
    - name: meshpeer
      port: 6783
      protocol: TCP
      targetPort: 6783
{{- end }}
  selector:
    {{- include "foo.matchLabels" . | nindent 4 }}
{{- end }}

# to
{{- with .Values.foo.a.super.super.long.miles.away }}
apiVersion: v1
kind: Service
metadata:
  {{- with .annotations }}
  annotations:
    {{ toYaml . | nindent 4 }}
  {{- end }}
  labels:
    {{- include "foo.labels" $ | nindent 4 }}
    {{- with .labels }}
    {{- toYaml . | nindent 4 }}
    {{- end }}
  name: {{ template "foo.fullname" $ }}-headless
  {{- include "foo.namespace" $ | indent 2 }}
spec:
  clusterIP: None
  ports:
    - name: http
      port: {{ .servicePort }}
      protocol: TCP
      targetPort: 9093
    {{- if .enableMeshPeer }}
    - name: meshpeer
      port: 6783
      protocol: TCP
      targetPort: 6783
    {{- end }}
  selector:
    {{- include "foo.matchLabels" $ | nindent 4 }}
{{- end }}
{{- end }}
```
- Flipping unnecessary condition `if` statement.
```
# from
{{- if (not condition) }}
statement_0
{{- else }}
statement_1

# to
{{- if condition }}
statement_1
{{- else }}
statement_0
```
- Adapt [k8s default labels](https://kubernetes.io/docs/concepts/overview/working-with-objects/common-labels/).

#### Checklist
- [x] [DCO](https://github.com/prometheus-community/helm-charts/blob/main/CONTRIBUTING.md#sign-off-your-work) signed
- [x] Chart Version bumped
- [x] Title of the PR starts with chart name

### CODE REVIEW

**Review by monotek — COMMENTED** [review]
imho we should keep the if clauses in a single line.
we don't use multiline format in any of the other charts and i also don't have seen this anywhere else before.

**Review by monotek — COMMENTED** [review]
If major refactor is done, we could also change the following:

* use statefulset deployment by default
  * upgrade instructions would be needed in readme
* change labels to meet current [kubernetes defaults](https://kubernetes.io/docs/concepts/overview/working-with-objects/common-labels/): 
  * upgrade instructions would be needed in readme
  * `helm create prometheus` should provide good starting point to get needed changes in _helpers.tpl 
  
But before your start with this one of the chart maintainers might wan't to add his view on this suggestions?

**Comment by zanac1986** [comment]
> If major refactor is done, we could also change the following:
> 
> * use statefulset deployment by default
>   
>   * upgrade instrcutions would be needed in readme
> * change labels to meet current [kubernetes defaults](https://kubernetes.io/docs/concepts/overview/working-with-objects/common-labels/):
>   
>   * upgrade instrcutions would be needed in readme
>   * `helm create prometheus` should provide good starting point to get needed changes
>   * in _helpers.tpl
> 
> But before your start with this one of the chart maintainers might wan't to add his view on this suggestions?

@Xtigyro @naseemkullah Thought?

File: charts/prometheus/Chart.yaml
Comment (monotek) [inline charts/prometheus/Chart.yaml:4]: Because of label change its a major version change.

```suggestion
version: 16.0.0
```

**Review by monotek — CHANGES_REQUESTED** [review]
Please add info about label change to readme.

Like here: https://github.com/prometheus-community/helm-charts/tree/main/charts/prometheus-node-exporter#3x-to-4x

**Comment by naseemkullah** [comment]
> > If major refactor is done, we could also change the following:
> 
> > 
> 
> > * use statefulset deployment by default
> 
> >   
> 
> >   * upgrade instructions would be needed in readme
> 
> > * change labels to meet current [kubernetes defaults](https://kubernetes.io/docs/concepts/overview/working-with-objects/common-labels/):
> 
> >   
> 
> >   * upgrade instructions would be needed in readme
> 
> >   * `helm create prometheus` should provide good starting point to get needed changes in _helpers.tpl
> 
> > 
> 
> > But before your start with this one of the chart maintainers might wan't to add his view on this suggestions?
> 
> 
> 
> @Xtigyro @naseemkullah Thought? 

I think the major release / breaking change should focus on using the alert manager, node exporter and push gateway charts as sub charts rather than the embedded components in this chart. It would deduplicate effort such as cleaning up syntax errors to just one place. My two cents. @monotek has been doing the heavy lifting for a while so I leave it up to him. In any case thank you @zanac1986 for this PR!

**Comment by monotek** [comment]
> I think the major release / breaking change should focus on using the alert manager, node exporter and push gateway charts as sub charts 

absolutly suppporting that! :)

**Pull request closed without merging**
